source/uri.c

```diff
@@ -0,0 +1,380 @@
+    parser->state = ON_AUTHORITY;
+}
+
+static const char *s_default_path = "/";
+
+static void s_parse_authority(struct uri_parser *parser, struct aws_byte_cursor *str) {
+    uint8_t *location_of_slash = memchr(str->ptr, '/', str->len);
+    uint8_t *location_of_qmark = memchr(str->ptr, '?', str->len);
+
+    if (!location_of_slash && !location_of_qmark && str->len) {
+        parser->uri->authority.ptr = str->ptr;
+        parser->uri->authority.len = str->len;
+
+        parser->uri->path.ptr = (uint8_t *)s_default_path;
+        parser->uri->path.len = 1;
+        parser->uri->request_uri = parser->uri->path;
+        parser->state = FINISHED;
+        aws_byte_cursor_advance(str, parser->uri->authority.len);
+    } else if (!str->len) {
+        parser->state = ERROR;
+        aws_raise_error(AWS_ERROR_MALFORMED_INPUT_STRING);
+        return;
+    }
+
+    uint8_t *end = str->ptr + str->len;
+    if (location_of_slash) {
+        parser->state = ON_PATH;
+        end = location_of_slash;
+    } else if (location_of_qmark) {
+        parser->state = ON_QUERY_STRING;
+        end = location_of_qmark;
+    }
+
+    parser->uri->authority.ptr = str->ptr;
+    parser->uri->authority.len = end - str->ptr;
+
+    aws_byte_cursor_advance(str, parser->uri->authority.len);
+
+    struct aws_byte_cursor authority_parse_csr = parser->uri->authority;
+
+    uint8_t *port_delim = memchr(authority_parse_csr.ptr, ':', authority_parse_csr.len);
+
+    if (!port_delim) {
+        parser->uri->port = 0;
+        parser->uri->host_name = parser->uri->authority;
+        return;
+    }
+
+    parser->uri->host_name.ptr = authority_parse_csr.ptr;
+    parser->uri->host_name.len = port_delim - authority_parse_csr.ptr;
+
+    size_t port_len = parser->uri->authority.len - parser->uri->host_name.len - 1;
+    port_delim += 1;
+    for (size_t i = 0; i < port_len; ++i) {
+        if (!isdigit(port_delim[i])) {
+            parser->state = ERROR;
+            aws_raise_error(AWS_ERROR_MALFORMED_INPUT_STRING);
+            return;
+        }
+    }
+
+    if (port_len > 5) {
+        parser->state = ERROR;
+        aws_raise_error(AWS_ERROR_MALFORMED_INPUT_STRING);
+        return;
+    }
+
+    /* why 6? because the port is a 16-bit unsigned integer*/
+    char atoi_buf[6] = {0};
+    memcpy(atoi_buf, port_delim, port_len);
+    parser->uri->port = (uint16_t)atoi(atoi_buf);
+    return;
+}
+
+static void s_parse_path(struct uri_parser *parser, struct aws_byte_cursor *str) {
+    parser->uri->request_uri = *str;
+
+    uint8_t *location_of_q_mark = memchr(str->ptr, '?', str->len);
+
+    if (!location_of_q_mark) {
+        parser->uri->path.ptr = str->ptr;
+        parser->uri->path.len = str->len;
+        parser->state = FINISHED;
+        aws_byte_cursor_advance(str, parser->uri->path.len);
+        return;
+    }
+
+    if (!str->len) {
+        parser->state = ERROR;
+        aws_raise_error(AWS_ERROR_MALFORMED_INPUT_STRING);
```

Comment:
Isn't this more an internal state machine error rather than input string?

Reply:
we only get to a bad state if the input string was malformed.